pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,758 @@
+                     */
+                    if (!taskQueue.isEmpty()) {
+                        executeTask();
+                        return CompletableFuture.completedFuture(null);
+                    } else if (STATE_UPDATER.compareAndSet(this, OperationState.None, OperationState.Operating)) {
+                        return task.get().whenComplete((ignore, throwable) -> {
+                            if (throwable != null && log.isDebugEnabled()) {
+                                log.debug("[{}] Failed to update index snapshot", topic.getName(), throwable);
+                            }
+                            STATE_UPDATER.compareAndSet(this, OperationState.Operating, OperationState.None);
+                        });
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                }
+                /*
+                  Only the operations of WriteSegment and DeleteSegment will be appended into the taskQueue.
+                  The operation will be canceled when the worker is close which means the topic is deleted.
+                 */
+                case WriteSegment, DeleteSegment -> {
+                    if (!STATE_UPDATER.get(this).equals(OperationState.Closed)) {
+                        taskQueue.add(new MutablePair<>(operationType, new MutablePair<>(taskExecutedResult, task)));
+                        executeTask();
+                        return taskExecutedResult;
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                }
+                case Clear -> {
+                    /*
+                      Do not clear the snapshots if the topic is used.
+                      If the users want to delete a topic, they should stop the usage of the topic.
+                     */
+                    if (STATE_UPDATER.compareAndSet(this, OperationState.None, OperationState.Closed)) {
+                        taskQueue.forEach(pair ->
+                                pair.getRight().getRight().get().completeExceptionally(
+                                        new BrokerServiceException.ServiceUnitNotReadyException(
+                                                String.format("Cancel the operation [%s] due to the"
+                                                                + " transaction buffer of the topic[%s] already closed",
+                                                        pair.getLeft().name(), this.topic.getName()))));
+                        taskQueue.clear();
+                        /*
+                          The task of clear all snapshot segments and indexes is executed immediately.
+                         */
+                        return task.get();
+                    } else {
+                        return FutureUtil.failedFuture(
```

Comment:
two questions:
- when this line is executed, the state of the topic will already be closed or fenced, right?
- All type of adding task is in the lock `synchronized (TopicTransactionBuffer.this)` except clear, right?

Reply:
When this line is executed, it means the state failed to change to closed. 
This close operation will be canceled, so the task in the queue should not be clear.